src/net_processing.h

```diff
@@ -84,7 +84,7 @@ class PeerManager : public CValidationInterface, public NetEventsInterface
 
     /** Process a single message from a peer. Public for fuzz testing */
     virtual void ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRecv,
-                                const std::chrono::microseconds time_received, const std::atomic<bool>& interruptMsgProc) = 0;
+                                const std::chrono::microseconds time_received, const std::atomic<bool>& interruptMsgProc) EXCLUSIVE_LOCKS_REQUIRED(g_mutex_msgproc_thread) = 0;
```

Comment:
Runtime lock assertions on the definitions for the methods with added annotations on their declarations, e.g. for the first commit (same for the later commits):

```diff
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -3135,6 +3135,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
                                      const std::chrono::microseconds time_received,
                                      const std::atomic<bool>& interruptMsgProc)
 {
+    AssertLockHeld(g_mutex_msgproc_thread);
 
@@ -4748,6 +4749,7 @@ bool PeerManagerImpl::MaybeDiscourageAndDisconnect(CNode& pnode, Peer& peer)
 
 bool PeerManagerImpl::ProcessMessages(CNode* pfrom, std::atomic<bool>& interruptMsgProc)
 {
+    AssertLockHeld(g_mutex_msgproc_thread);
 
@@ -5240,6 +5242,7 @@ bool PeerManagerImpl::SetupAddressRelay(const CNode& node, Peer& peer)
 
 bool PeerManagerImpl::SendMessages(CNode* pto)
 {
+    AssertLockHeld(g_mutex_msgproc_thread);
```

Reply:
Added the runtime assertions for these three, but not any of the private/internal methods in `net_processing`, since that doesn't seem likely to have any benefit.

Reply:
Actually, the guidelines are pretty explicit for this:

>   - Combine annotations in function declarations with run-time asserts in
    function definitions:

I think it is better to remove or modify a rule from the guidelines instead of violating it.